[vertex tool] fix slow deletion of many vertices

ViperMiniQ opened the pull request:
refs: #58536

This PR adds a new function `deleteVertices` to avoid constant calls to `detach()` (which clones the geometry) which cooks the CPU when a large number of vertices are to be deleted. During my debugging, I didn't notice reference count to geometry being less than 2 at any time, so `detach()` is always being called.

Debug information is added to the function on fails.

In order to avoid code duplication and easier future fixes/additions, I suggest to modify `deleteVertex` to add a vertex to a list and internally call `deleteVertices`. Waiting for opinions on this.

ViperMiniQ committed:
add deleteVertices and call it from the vertex tool instead of single calls to deleteVertex

nyalldawson commented on src/core/geometry/qgsgeometry.cpp:
```diff
@@ -678,6 +678,64 @@ bool QgsGeometry::deleteVertex( int atVertex )
+  QList<int> sortedVertices = vertices;
+  std::sort( sortedVertices.begin(), sortedVertices.end(), std::greater<int>() );
+
+  // maintain compatibility with < 2.10 API
+  if ( QgsWkbTypes::flatType( d->geometry->wkbType() ) == Qgis::WkbType::MultiPoint )
+  {
+    detach();
+    QgsGeometryCollection *collection = static_cast< QgsGeometryCollection * >( d->geometry.get() );
+    
+    for ( int vertex : sortedVertices )
+    {
+      if ( !collection->removeGeometry( vertex ) )
+        return false;
+    }
+    return true;
+  }
```
This isn't necessary, as it's a completely new API

uclaros commented:
I believe the return values should be revised, and this is something that should be apparent when writing tests for this.
Let me explain.
Since the vertex indices are sorted in descending order there are the following cases where `deleteVertices` may return false:
- max index provided is bigger than total number of vertices.
  - In this case we correctly return false, as the geometry is unchanged
- min index provided is negative.
  - In this case we return false, regardless of how many vertices were actually removed. I believe that we should ignore negative values and return true if at least one vertex was actually removed.
- Index provided was implicitly removed by removal of its adjacent vertex
  - In this case we should ignore the missing vertex and not stop iteration, eventually returning `true`

An example for the last case is:
```python
>>>l=QgsGeometry.fromWkt('LINESTRING( 0 0, 42 0)')
>>>l
<QgsGeometry: LineString (0 0, 42 0)>
>>>l.deleteVertex(1)
True
>>>l
<QgsGeometry: LineString EMPTY>
```
The `0` vertex gets automatically removed as we can't have a single vertex linestring.
In this case `l.deleteVertices([0,1])` would return `false` and that would be quite confusing.

ViperMiniQ committed:
remove old code

ViperMiniQ committed:
check if vertex is still valid for deletion, reset original geometry on fail

ViperMiniQ commented:
@uclaros 

I disagree with allowing negative values (or any non-existing values). 
The function should return `true` only if all specified vertices were successfully deleted, similar to how `deleteVertex` returns `false` when a non-existing vertex is passed. 
It should be the user's responsibility to avoid passing invalid indexes. Internally, if garbage values are provided, the function shouldn't delete only some vertices and silently ignore the others.

> 
> Index provided was implicitly removed by removal of its adjacent vertex
> 
> In this case we should ignore the missing vertex and not stop iteration, eventually returning true

Thank you for pointing this out. I’ve added a check to ensure a vertex hasn’t already been deleted during the operation.

Also, I’ve added logic to copy the original geometry so it can be restored in case of failure. 
This prevents ending up with partially edited geometry.

I will see to add the tests tomorrow.

uclaros commented:
I'm easy about failing on negative values. My main point was not to delete vertices and then return false, which is fixed with your latest changes.

ViperMiniQ committed:
handle rings

ViperMiniQ committed:
add tests

github-actions[bot] committed:
auto-fix pre-commit issues

uclaros commented:
It looks good, deleting the starting/closing vertex is tested, as well as deleting a whole part.
Maybe adding a test for deleting part of a circular string too?

ViperMiniQ commented:
I don't think it is possible doing it this way (I cannot solve `CompoundCuve`), most likely the only option is to actually implement `deleteVertices` for each geometry type.
